Connect contacts through script pages, hook up a done page at the end

nickoneill opened the pull request:
Plumbed the real contacts through the script pages, added a placeholder done for now

cselin commented on FiveCalls/FiveCalls/IssueContactDetail.swift:
```diff
@@ -50,19 +50,38 @@ struct IssueContactDetail: View {
+    }
+    
     func nextButtonTitle() -> String {
         return "Next contact"
```
localize

cselin commented on FiveCalls/FiveCalls/IssueContactDetail.swift:
```diff
@@ -50,19 +50,38 @@ struct IssueContactDetail: View {
+        return Array(remainingContacts.dropFirst())
+    }
+    
     func nextButtonTitle() -> String {
```
this can probably just be replaced with localized string inline

cselin commented on FiveCalls/FiveCalls/IssueDetail.swift:
```diff
@@ -26,15 +28,15 @@ struct IssueDetail: View {
                 }.padding(.bottom, 16)
-                NavigationLink(destination: IssueContactDetail(issue: issue, contact: .housePreviewContact)) {
+                NavigationLink(destination: IssueContactDetail(issue: issue, remainingContacts: issue.contactsForIssue(allContacts: store.state.contacts))) {
                     PrimaryButton(title: "See your script", systemImageName: "megaphone.fill")
```
localize

nickoneill commented on FiveCalls/FiveCalls/Contact.swift:
```diff
@@ -98,6 +98,27 @@ extension Contact {
     }
+    
+    // this has some overlap with other area -> string conversions but I haven't thought about it long enough to combine them
+    func officeDescription() -> String {
```
👍

nickoneill commented on FiveCalls/FiveCalls/IssueContactDetail.swift:
```diff
@@ -50,19 +50,38 @@ struct IssueContactDetail: View {
         }
     }
     
+    func currentContact() -> Contact {
```
No, they fit better as properties, changed.

nickoneill commented:
@cselin addressed your comments and added a router + better applied the navigation stack concept. I'm not sure I like the navigation models, I was trying to do the same with just tuples but I suppose those wouldn't be hashable (and I couldn't figure out the right type signature). Still, I don't dislike the concept of having all the nav destinations at the root stack.

cselin commented on FiveCalls/FiveCalls/IssueDetail.swift:
```diff
@@ -38,11 +38,12 @@ struct IssueDetail: View {
                 }.padding(.bottom, 16)
-                NavigationLink(destination: IssueContactDetail(issue: issue, remainingContacts: issue.contactsForIssue(allContacts: store.state.contacts))) {
-                    PrimaryButton(title: "See your script", systemImageName: "megaphone.fill")
+                NavigationLink(value: IssueDetailNavModel(issue: issue, contacts: issue.contactsForIssue(allContacts: store.state.contacts))) {
```
nit: I wonder if we need to pass the main app store into this view, it looks like we're only using it to access the contacts, but not performing any actions. We could probably just pass in the contacts to this view

nickoneill replied:
I don't quite understand; issue detail only gets an issue and an array of the remaining contacts, they are filtered by the `contactsForIssue` method which matches on area.

cselin replied:
Right, that's what I mean, we include the main store on line 12 of this file, but if we're only using it to get the contacts from the state, we could just include pass the contacts into the initializer for this view, unless you think we'll need access to the full store in the future. Just being overly picky here if we wanted to limit access is all.

nickoneill replied:
ok, I get it. Yeah, I'll give that a try

cselin commented on FiveCalls/FiveCalls/IssueDone.swift:
```diff
@@ -9,13 +9,17 @@
                 Text("Issue \(issue.name) done page")
-                Button(action: {}, label: {
+                Button(action: {
+                    router.backToRoot()
```
nit: this could be done via the unidirectional store instead if we wanted to keep everything in that flow. We could add an action `backToRoot` (handled in middleware) that could be sent from anywhere that has access to the store without having to expose the routing details.

nickoneill replied:
I'm going to give this a shot to see how it feels but I'll do it in another PR

cselin commented on FiveCalls/FiveCalls/Router.swift:
```diff
@@ -0,0 +1,53 @@
+    }
+}
+
+struct IssueDetailNavModel {
```
nit: these probably belong in their own file, or maybe in the view files that use them?

nickoneill replied:
👍 will move these into the views that use that model

cselin commented on FiveCalls/FiveCalls/Router.swift:
```diff
@@ -0,0 +1,53 @@
+
+import SwiftUI
+
+class Router: ObservableObject {
```
love this!

cselin commented:
> @cselin addressed your comments and added a router + better applied the navigation stack concept. I'm not sure I like the navigation models, I was trying to do the same with just tuples but I suppose those wouldn't be hashable (and I couldn't figure out the right type signature). Still, I don't dislike the concept of having all the nav destinations at the root stack.

Love the router. wrt the nav models, we did something similar at my last company, though we called them view state and embedded them in the views. They were similar in concept to view models that represented a subset of the state they were derived from that we just needed in a respective view. This was all pre-navigation stack so it's interesting to see what must be done to keep it hashable. I think what you've done here works well.

cselin commented on FiveCalls/FiveCalls/Dashboard.swift:
```diff
@@ -10,13 +10,14 @@ import SwiftUI
 
 struct Dashboard: View {
     @EnvironmentObject var store: Store
+    @EnvironmentObject var router: Router
```
wondering if this particular router belongs directly on this view only, rather than as an environmental object. Since popping to root will only ever go back to the dashboard it seems like it could be restricted to here. Just a thought.

nickoneill replied:
Are you suggesting this in conjunction with moving router actions into middleware? Intermediate screens currently need to know about the router as well so they can pop a single item off the path to go back.

cselin replied:
Not necessarily tied to router actions in the middleware. Just thinking that this router could be an @StateObject declared and initialized here if the router is only used in the Dashboard and subviews. Arguably, the same could be said for the store since everything in it applies to the dashboard and subviews, so it's probably fine to keep both declared and initialized one level up in the app. You can disregard this thread ^

cselin commented on FiveCalls/FiveCalls/AppDelegate.swift:
```diff
@@ -39,7 +39,8 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
         if USE_NEW_SWIFTUI_INTERFACE {
             let store = Store(state: AppState(), middlewares: [appMiddleware()])
-            window?.rootViewController = UIHostingController(rootView: Dashboard().environmentObject(store))
+            let router = Router()
```
see my comments elsewhere about 1) restricting this dashboard specific router to the dashboard and subviews and 2) hiding this behind the store so consumers don't need to know the internals about how routing works.

cselin reviewed (approved):
Looks good. I proposed some ideas about adding a bit more separation for the new router for consideration.
